prover/crates/bin/witness_generator/src/leaf_aggregation.rs

```diff
@@ -265,19 +265,93 @@ pub async fn prepare_leaf_aggregation_job(
     skip_all,
     fields(l1_batch = %job.block_number, circuit_id = %job.circuit_id)
 )]
-pub fn process_leaf_aggregation_job(
+pub async fn process_leaf_aggregation_job(
     started_at: Instant,
     job: LeafAggregationWitnessGeneratorJob,
+    object_store: Arc<dyn ObjectStore>,
+    max_circuits_in_flight: usize,
 ) -> LeafAggregationArtifacts {
     let circuit_id = job.circuit_id;
-    let subsets = (
-        circuit_id as u64,
-        job.closed_form_inputs.1,
-        job.closed_form_inputs.0,
-    );
-    let leaf_params = (circuit_id, job.leaf_params);
-    let (aggregations, closed_form_inputs) =
-        create_leaf_witnesses(subsets, job.proofs, job.base_vk, leaf_params);
+    let queues = split_recursion_queue(job.closed_form_inputs.1);
+
+    let aggregations = queues
+        .iter()
+        .cloned()
+        .map(|queue| (circuit_id as u64, queue))
+        .collect();
+
+    let mut proof_ids_iter = job.proofs_ids.into_iter();
+    let mut proofs_ids = vec![];
+    for queue in queues.iter() {
+        let proofs_ids_for_queue: Vec<_> = (&mut proof_ids_iter)
+            .take(queue.num_items as usize)
+            .collect();
+        assert_eq!(queue.num_items as usize, proofs_ids_for_queue.len());
+        proofs_ids.push(proofs_ids_for_queue);
+    }
+
+    let semaphore = Arc::new(Semaphore::new(max_circuits_in_flight));
+
+    let mut handles = vec![];
+    for (circuit_idx, (queue, proofs_ids_for_queue)) in
+        queues.into_iter().zip(proofs_ids).enumerate()
+    {
+        let semaphore = semaphore.clone();
+
+        let object_store = object_store.clone();
+        let queue = queue.clone();
+        let base_vk = job.base_vk.clone();
+        let leaf_params = (circuit_id, job.leaf_params.clone());
+
+        let handle = tokio::task::spawn(async move {
+            let _permit = semaphore
+                .acquire()
+                .await
+                .expect("failed to get permit to process queues chunk");
+
+            let proofs = load_proofs_for_job_ids(&proofs_ids_for_queue, &*object_store).await;
+            let base_proofs = proofs
+                .into_iter()
+                .map(|wrapper| match wrapper {
+                    FriProofWrapper::Base(base_proof) => base_proof,
+                    FriProofWrapper::Recursive(_) => {
+                        panic!(
+                            "Expected only base proofs for leaf agg {} {}",
+                            job.circuit_id, job.block_number
+                        );
+                    }
+                })
+                .collect();
+
+            let (_, circuit) = create_leaf_witness(
```

Comment:
Welp, sorry for raising it again, I should've communicated it better -- but it looks like this function is blocking, and it's in the `tokio::task::spawn` context; e.g. it'll be executed within runtime. I'm not sure if `split_recursion_queue` above is CPU heavy, but if it isn't, then the following changes can be done:

1. It's fine to `tokio::spawn` the `process_leaf_aggregation_job`. Same for `basic_circuits` -- I've checked and in fact we call `zkevm_test_harness::external_calls::run` from `spawn_blocking` context already.
2. Here, `spawn` should be replaced with `spawn_blocking`, and the handle should be passed to block on `save_recursive_layer_prover_input_artifacts`.

Reply:
Actually it is not CPU heavy at all

Reply:
I mean `create_leaf_witness` is very light too

Reply:
`save_recursive_layer_prover_input_artifacts` is just I/O too, should we block on it?

Reply:
Ok, both `split_recursion_queue` and `create_leaf_witness` are very lightweight. I don't think it will cause any problems.